### PR TITLE
Fix #3860, tearing down TCP connection for send_request_cgi

### LINKED ISSUE #3860: send_request_cgi doesn't tear down the TCP connection
*This issue was [RM8802](https://dev.metasploit.com/redmine/issues/8802), originally filed by @Meatballs1*

I have tried doing `send_request_cgi; disconnect` and `send_request_cgi; cleanup` from a `Msf::Exploit::HTTP::Client` aux module but it doesn't close down. If you reload the module it will tear down the connection.

For things like `auxiliary/scanner/sap/sap_icm_urlscan` which is bruteforcing URLs this causes loads of established connections to remain (`netstat -an |grep CLOSE_WAIT`)

Quoted from IRC:

````
<@hdm> aha
---» le_tropico (~le_tropic@194.44.205.54) has Joined #metasploit
<@hdm> disconnect doesnt close the socket unless its using the "non-global" socket
<@hdm> try calling self.client.close instead of disconnect
<@Meatballs> ah ok will give it a try, I dont really understand global socket/non, did see some of that code :)
<@hdm> not clear why self.client shouldnt be closed autoamtically during disconnect()
<@egypt> or `mysock = connect` and then `disconnect(mysock)` ?
<@Meatballs> I guess if its been sent to a handler?
<@egypt> for findsock paylods, no?
<@hdm> every module/mixin that does socket stuff tends to save off a single socket as the "main" socket, either self.sock/self.client, etc
<@egypt> (which are all but dead but that's beside the point)
<@hdm> it lets you shortcut send/recv(socket) as just connect/send/recv/close
<@hdm> findsock is probably the reason for it
````

### PR DESCRIPTION
**Pull request by wchen-r7**
# What This Patch Does

This patch fixes an issue explained in #3860.

When `send_request_cgi` is used without the global socket, it does not close the TCP connection even when the `disconnect` method is called explicitly. It also does not close when the module is completed. This patch addresses those problems.

The patch also ensures `super` is called in the scanner mixin's `cleanup` method, that way auxiliary modules that use the scanner and HttpClient mixin will be able to close the TCP connection properly.

Fix #3860 
MS-440
# Verification
- [x] `cd` to a directory where you prefer to host a fake web server.
- [x] `echo "Hello world" > test.html`. The test script will be requesting this.
- [x] Under the same directory, do: `ruby -run -e httpd . -p 8181` to start a web server
- [x] Save the following script so you can use in msfconsole:

``` ruby
require 'msf/core'

class MetasploitModule < Msf::Auxiliary

  include Msf::Exploit::Remote::HttpClient
  include Msf::Auxiliary::Scanner

  def initialize(info = {})
    super(update_info(info,
      'Name'           => 'Test',
      'Description'    => %q{
        Test
      },
      'Author'         => [ 'sinn3r' ],
      'License'        => MSF_LICENSE,
      'References'     => 
        [
          ['URL', 'http://metasploit.com']
        ]
    ))
  end

  def print_status(msg='')
    super("#{peer} - #{msg}")
  end

  def test_case_1
    print_status("Sending HTTP request...")
    res = send_request_cgi({
      'uri' => '/test.html',
      'method' => 'GET'
    })

    if res
      print_status(res.body.strip)
    end

    print_status('Sleeping for 10 seconds.')
    print_status('Do netstat, and you should see port 8181 is connected')
    sleep(10)

    print_status('Manually disconnecting TCP connection now...')
    disconnect
    print_status('Do netstat, you should see that port 8181 is no longer connected')
    sleep(10)
  end

  def test_case_2
    print_status("Sending HTTP request...")
    res = send_request_cgi({
      'uri'    => '/test.html',
      'method' => 'GET',
      'global' => true
    })

    if res
      print_status(res.body.strip)
    end

    print_status('Sleeping for 10 seconds.')
    print_status('Do netstat, and you should see port 8181 is connected')
    sleep(10)

    disconnect
    print_status('Do netstat, you should see that port 8181 is no longer connected')
    sleep(10)
  end

  def test_case_3
    print_status("Sending HTTP request...")
    send_request_cgi({
      'uri'    => '/test.html',
      'method' => 'GET'
    })

    print_status('The module is about to end. After it finishes, you should not see any port 8181 connected.')
  end

  def run_host(ip)
    test_case_1
    print_line
    test_case_2
    print_line
    test_case_3
  end

end
```
- [x] Start msfconsole
- [x] Do: `use [path to the test module]`
- [x] Do: `set RHOST [IP]`
- [x] Do: `set RPORT8181`
- [x] Follow the instructions the module gives you. Basically it just tells you when to netstat to check if you're connected to port 8181 or not.

Since this change could potential break a lot of modules, please feel free to explore other edge cases to test.

### CODE REVIEW

**Comment by wchen-r7** [comment]
@bcook-r7 is currently reviewing so I'll assign to him.

**Comment by bcook-r7** [comment]
This doesn't change the return value of send_request_cgi, so it addresses the concern that Juan originally had. Works as advertised, I'll poke at this a bit more and hopefully land soon.

**Comment by bcook-r7** [comment]
Thanks @wchen-r7

**Comment by tdoan-r7** [comment]
@bcook-r7 @wchen-r7 can you add release notes please?

**Comment by wchen-r7** [comment]
# Release Notes

When send_request_cgi is used without the global socket, it does not close the TCP connection even when the disconnect method is called explicitly. It also does not close when the module is completed. This patch addresses those problems.

File: lib/msf/core/exploit/http/client.rb [hunk 1]
```diff
@@ -297,6 +297,10 @@ def disconnect(nclient = self.client)
     end
 
     if (nclient == self.client)
+      if self.client.respond_to?(:close)
```
Comment (wchen-r7) [inline lib/msf/core/exploit/http/client.rb]: @wchen-r7 yo look at this again fool
